src/lib/bundles/bundles.ts

```diff
@@ -0,0 +1,308 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Transform } from 'node:stream';
+import type { Readable } from 'node:stream';
+import { createHash } from 'node:crypto';
+import {
+  MIN_BINARY_SIZE,
+  byteArrayToLong,
+  deepHash,
+  indexToType,
+} from 'arbundles';
+import { deserializeTags } from './tags.js';
+
+export interface SignatureMeta {
+  sigLength: number;
+  pubLength: number;
+  sigName: string;
+}
+
+export enum SignatureConfig {
+  ARWEAVE = 1,
+  ED25519,
+  ETHEREUM,
+  SOLANA,
+  INJECTEDAPTOS = 5,
+  MULTIAPTOS = 6,
+  TYPEDETHEREUM = 7,
+}
+
+export const SIG_CONFIG: Record<SignatureConfig, SignatureMeta> = {
+  [SignatureConfig.ARWEAVE]: {
+    sigLength: 512,
+    pubLength: 512,
+    sigName: 'arweave',
+  },
+  [SignatureConfig.ED25519]: {
+    sigLength: 64,
+    pubLength: 32,
+    sigName: 'ed25519',
+  },
+  [SignatureConfig.ETHEREUM]: {
+    sigLength: 65,
+    pubLength: 65,
+    sigName: 'ethereum',
+  },
+  [SignatureConfig.SOLANA]: {
+    sigLength: 64,
+    pubLength: 32,
+    sigName: 'solana',
+  },
+  [SignatureConfig.INJECTEDAPTOS]: {
+    sigLength: 64,
+    pubLength: 32,
+    sigName: 'injectedAptos',
+  },
+  [SignatureConfig.MULTIAPTOS]: {
+    sigLength: 64 * 32 + 4, // max 32 64 byte signatures, +4 for 32-bit bitmap
+    pubLength: 32 * 32 + 1, // max 64 32 byte keys, +1 for 8-bit threshold value
+    sigName: 'multiAptos',
+  },
+  [SignatureConfig.TYPEDETHEREUM]: {
+    sigLength: 65,
+    pubLength: 42,
+    sigName: 'typedEthereum',
+  },
+};
+
+const isValidSignatureConfig = (value: number): value is SignatureConfig => {
+  return Object.values(SignatureConfig).includes(value);
+};
+
+const getSignatureMeta = (signatureType: number): SignatureMeta => {
+  if (isValidSignatureConfig(signatureType)) {
+    return SIG_CONFIG[signatureType];
+  } else {
+    throw new Error('Invalid signature type');
+  }
+};
+
+export interface DataItemInfo {
+  anchor: string;
+  dataOffset: number;
+  dataSize: number;
+  id: string;
+  offset: number;
+  owner: string;
+  ownerOffset: number;
+  ownerSize: number;
+  sigName: string;
+  signature: string;
+  signatureOffset: number;
+  signatureSize: number;
+  signatureType: number;
+  size: number;
+  tags: { name: string; value: string }[];
+  target: string;
+}
+
+export const processBundleStream = async (
```

Comment:
It'll take a little extra time, but given we'll like extract this into a library soon, I think we go ahead and add unit tests for it.

Reply:
I agree 👍